### PR TITLE
[mlir][vector] Fix invalid IR in `RewriteBitCastOfTruncI`

### PR DESCRIPTION
**Pull request by matthias-springer**
This commit fixes `Dialect/Vector/vector-rewrite-narrow-types.mlir` when running with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`.

```
within split at llvm-project/mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:1 offset :118:8: error: 'arith.trunci' op operand type 'vector<3xi16>' and result type 'vector<3xi16>' are cast incompatible
  %1 = vector.bitcast %0 : vector<16xi3> to vector<3xi16>
       ^
within split at llvm-project/mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:1 offset :118:8: note: see current operation: %48 = "arith.trunci"(%47) : (vector<3xi16>) -> vector<3xi16>
LLVM ERROR: IR failed to verify after pattern application
```

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 1]
```diff
@@ -664,8 +664,8 @@ struct BitCastRewriter {
 
 } // namespace
 
-[[maybe_unused]] static raw_ostream &operator<<(raw_ostream &os,
-                               const SmallVector<SourceElementRangeList> &vec) {
+[[maybe_unused]] static raw_ostream &
+operator<<(raw_ostream &os, const SmallVector<SourceElementRangeList> &vec) {
```
Comment (hanhanW) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:668]: unrelated changes?

**Review by hanhanW — APPROVED** [review]
Ugh, I've hit similar issues before. This is a bit annoying, thanks for fixing it. Will we have CI running with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`?

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
What implications does this have on tests ?
Also, this is perfectly valid in LLVM, this suggests to me the "CastCompatible" verifier in MLIR may not a good fit here.

**Comment by joker-eph** [comment]
> Also, this is perfectly valid in LLVM

Which part?
As far as I can tell, the MLIR error is for:

> %48 = "arith.trunci"(%47) : (vector<3xi16>) -> vector<3xi16>

And looking at some equivalent in LLVM I see:

>error: invalid cast opcode for cast from 'i16' to 'i16'
>  %r8 = trunc i16 undef to i16

or: 

> error: invalid cast opcode for cast from '<3 x i16>' to '<3 x i16>'
>  %r8 = trunc <3 x i16> undef to <3 x i16>

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp (hunk 1 above)
Comment (matthias-springer) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:668]: looks like `clang-format` did this

**Review by nicolasvasilache — APPROVED** [review]
@joker-eph ah thanks for digging deeper, I was under the impression we were disabling `16xi3 -> 3xi16` type bitcasts.

Seems we are just removing noops then, which seems fine.

I can live with this, although it would seem that this would be better as a folder in LLVM when certain ops are noops ..

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp (hunk 1 above)
Comment (hanhanW) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:668]: okay, that's probably fine.

**Comment by matthias-springer** [comment]
> Ugh, I've hit similar issues before. This is a bit annoying, thanks for fixing it. Will we have CI running with `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`?

Yes, that's the plan. There are a few more tests that I have to fix before we can enable it. I also have to improve the running time of the expensive checks, right now the expensive checks are pretty slow.
